### ISSUE TITLE
xrdp_wm_log_msg: Error connecting to sesman: 127.0.0.1 port: 3350

### ISSUE THREAD

**Issue by min050820** [neutrinolabs/xrdp#1855]
I believe this issue is the same as #1777. I opened new issue because I'm okay with installing nmap.

My system is aarch64 (same as that issue), running ubuntu 20.04 with linux 4.9.
_Perhaps this issue has something to do with aarch64?_

xrdp-sesman service is running. I also tried restarting them.
My firewall also is okay. I don't have ufw, and tried flushing iptable.

Result of what @matt335672 suggested:
```
root@odroid:~# ps -e -o pid,netns,command | grep xrdp
   2760 4026531955 /usr/sbin/xrdp-sesman
   2860 4026531955 /usr/sbin/xrdp
  25022 4026531955 grep --color=auto xrdp

root@odroid:~# ss -alp 'sport = 3350'
RTNETLINK answers: Invalid argument
Netid      State       Recv-Q      Send-Q            Local Address:Port             Peer Address:Port      Process
tcp        LISTEN      0           2                         [::1]:3350                     [::]:*          users:(("xrdp-sesman",pid=2760,fd=7))

root@odroid:~# nmap -p 3350 127.0.0.1
Starting Nmap 7.80 ( https://nmap.org ) at 2021-04-08 15:24 KST
Nmap scan report for localhost (127.0.0.1)
Host is up (0.00011s latency).

PORT     STATE  SERVICE
3350/tcp closed findviatv

Nmap done: 1 IP address (1 host up) scanned in 0.17 seconds
```

xrdp.log:
```
[20210408-15:16:08] [INFO ] connected client computer name: DESKTOP-URBOSA
[20210408-15:16:08] [INFO ] adding channel item name rdpdr chan_id 1004 flags 0x80800000
[20210408-15:16:08] [INFO ] adding channel item name rdpsnd chan_id 1005 flags 0xc0000000
[20210408-15:16:08] [INFO ] adding channel item name cliprdr chan_id 1006 flags 0xc0a00000
[20210408-15:16:08] [INFO ] adding channel item name drdynvc chan_id 1007 flags 0xc0800000
[20210408-15:16:08] [INFO ] TLS connection established from ::ffff:218.158.49.161 port 51387: TLSv1.2 with cipher ECDHE-RSA-AES256-GCM-SHA384
[20210408-15:16:08] [DEBUG] xrdp_00006152_wm_login_mode_event_00000001
[20210408-15:16:08] [INFO ] Cannot find keymap file /etc/xrdp/km-e0010412.ini
[20210408-15:16:08] [INFO ] Loading keymap file /etc/xrdp/km-00000412.ini
[20210408-15:16:08] [WARN ] local keymap file for 0xe0010412 found and doesn't match built in keymap, using local keymap file
[20210408-15:16:17] [DEBUG] xrdp_wm_log_msg: connecting to sesman ip 127.0.0.1 port 3350
[20210408-15:16:21] [DEBUG] Closed socket 18 (AF_INET6 ::1 port 55414)
[20210408-15:16:25] [DEBUG] Closed socket 18 (AF_INET6 ::1 port 55416)
[20210408-15:16:29] [INFO ] Socket 12: AF_INET6 connection received from ::ffff:14.32.139.148 port 55412
[20210408-15:16:29] [DEBUG] Closed socket 12 (AF_INET6 ::ffff:218.158.49.162 port 3389)
[20210408-15:16:29] [DEBUG] Closed socket 11 (AF_INET6 :: port 3389)
[20210408-15:16:29] [DEBUG] TLSv1.3 enabled
[20210408-15:16:29] [DEBUG] TLSv1.2 enabled
[20210408-15:16:29] [DEBUG] Security layer: requested 3, selected 1
[20210408-15:16:29] [DEBUG] Closed socket 12 (AF_INET6 ::ffff:218.158.49.162 port 3389)
[20210408-15:16:29] [DEBUG] Closed socket 18 (AF_INET6 ::1 port 55418)
[20210408-15:16:33] [ERROR] xrdp_wm_log_msg: Error connecting to sesman: 127.0.0.1 port: 3350
[20210408-15:16:33] [DEBUG] Closed socket 18 (AF_INET6 ::1 port 55420)
[20210408-15:16:33] [DEBUG] return value from xrdp_mm_connect 1
```
(DESKTOP-URBOSA is my computer running Windows 10)

**Comment by matt335672** [comment]
Interesting.

I think this is another manifestation of #1805.

The listen/connect code in xrdp isn't great. There are moves afoot to replace this interface with a UNIX domain socket which will fix a lot of these niggles.

I think you're right in that it's something to do with aarch64, and in particular it's your kernel version.

Sequence of events is:-
- Despite the fact sesman is configured to listen on 127.0.0.1, it's listening on ::1. This happens when xrdp is configured with `--enable-ipv6`
- Later, xrdp logs it's trying to connect to port 3350 on 127.0.0.1 but instead tries to connect to port 3350 on ::1 - I've stepped through the code to confirm this. However, I think your kernel 4.9 is for some reason not succeeding on this, and that's where the holdup happens.

I can offer you two solutions, neither of which is ideal:-
1. Rebuild xrdp without `--enable-ipv6`
2. Set the `ListenAddress` in `/etc/xrdp/sesman.ini` to `0.0.0.0` rather than `127.0.0.1`. This means sesman is listening externally to port 3350 which isn't at all secure. If you go this route, I suggest you set up a firewall to prevent incoming connections on this port.

**Comment by min050820** [comment]
Oh. I didn't notice local address saying [::1] instead of 127.0.0.1 until you told so.

Actually, my xrdp setup was working before I updated my system -- which also updated `linux`. So I guess you're correct that my kernel is somehow broken. ip6tables don't have any rules, by the way.

However, making sesman listen on 0.0.0.0 did not solve the problem.

So I tried rebuilding xrdp with --disable-ipv6 and it worked! I'm closing this issue, but I'll be eager to test things if you make some fixes.

Just in case:
`$ uname -a`
`Linux odroid 4.9.241-113 #1 SMP PREEMPT Wed Mar 17 18:56:00 UTC 2021 aarch64 aarch64 aarch64 GNU/Linux`

**Comment by sainsg** [comment]
Hi Quartzsaber, hope you are well. I know this is closed but I have been struggling to get this working myself. Can you explain how you would go about rebuilding xrdp with the disable ipv6 flag as I have tried it and it does not appear to be a valid parameter I can use when I install. I'm running a fresh new install of Almalinux 8 trying to connect via xRDP to an XFCE session. Any help would be much appreciated !

**Comment by quartzsaber** [comment]
@sainsg I believe you should follow [Compiling section in README.md](https://github.com/neutrinolabs/xrdp#compiling). While following that, pass `--disable-ipv6` as a parameter for configure step and not the install step. (e.g. `./configure --disable-ipv6`)

Frankly, I don't remember how I did that. It was a year ago anyway. So, please don't ping me if it still doesn't work. I can't really help you. Plus, posting a message here will send a notification to the maintainer too. I'd rather not disturb them.

Just asking, but is your machine also an aarch64 one? If so, which board are you using? If you were using a completely different one than mine, that could mean it's not just my board's kernel being broken, but something being wrong in xrdp's side.

**Comment by nathan-gilbert** [comment]
I ran into this same issue with OpenSuSE Tumbleweed. The solution was to compile with `--disable-ipv6` and drop those binaries in `/usr/sbin` then restart the service `service xrdp restart` and `service xrdp-sesman restart`.

**Comment by matt335672** [comment]
This should be a thing of the past in the next major release of xrdp, now we've merged #2207.
